[Docs] Improvements to custom flavor and custom orchestrator pages (#1747)
zenml-io/zenml: docs/book/stacks-and-components/component-guide/writing-custom-components.md

---
description: How to extend ZenML with plug-in flavor components.
---

# Integrate ZenML with your tooling stack

> ZenML is built on the philosophy of composability and reusability. It allows users to develop custom component flavors for each type of pipeline component. This guide will help you understand what a flavor is, and how you can develop and register your own custom flavor in ZenML.

Reviewer: ```suggestion
ZenML is built on the philosophy of composability and reusability. It allows users to develop custom component flavors for each type of stack component. This guide will help you understand what a flavor is, and how you can develop and register your own custom flavor in ZenML.
```

## Understanding Component Flavors

In ZenML, a component type is a broad category that defines the functionality of a pipeline component. Each type can have multiple flavors, which are specific implementations of the component type. For instance, the type `orchestrator` can have flavors like `local`, `kubernetes`, etc. Each flavor is a unique implementation of the orchestrator functionality.

In order to create a custom component, one must create:

1. A custom flavor class (inheriting from `zenml.stack.Flavor` or any of its child classes)
2. An implementation class (inheriting from `zenml.stack.StackComponent` or any of its child classes)

> ## Writing a Custom Flavor Class

Reviewer: ```suggestion
## Writing a custom flavor class
```

## Registering a flavor

Once you have implemented your custom flavor and implementation class, you can register your component using the ZenML CLI. Use the `flavor register` command and provide the path to your flavor class (see further below for a concrete example):

```shell
# This points to the flavor class, not the implementation class
zenml <component_type> flavor register <path.to.MyFlavor>
```

Replace `<component_type>` with the type of component you're developing the flavor for, and `<path.to.MyFlavor>` with the path to your flavor class.

For example, if your custom orchestrator flavor class `MyOrchestratorFlavor` is defined in `flavors/my_flavor.py`, you'd register it by doing:

```shell
zenml orchestrator flavor register flavors.my_flavor.MyOrchestratorFlavor
```

### Listing Flavors

You can list all registered flavors for a component type using the `flavor list` command:

```shell
zenml <component_type> flavor list
```

Replace `<component_type>` with the type of component you're interested in, e.g. `orchestrator`.

### Understanding flavor registration

> When you register a flavor, ZenML imports the flavor class and stores its configuration in the ZenML repository. This allows ZenML to use your flavor when running pipelines, even if the dependencies required by the flavor are not installed in the current environment.

Reviewer 2: > stores its configuration in the ZenML repository

I'm not sure repository is the word here, maybe server/database?

> In ZenML, the config is defined in flavor class in the `def config_class(self) -> Type[StackComponentConfig]` method, while in the settings are defined in the implementation class in the `def settings_class(self) -> Type[StackComponentSettings]` function.

Reviewer 2: ```suggestion
In ZenML, the config is defined in flavor class in the `def config_class(self) -> Type[StackComponentConfig]` property, while in the settings are defined in the implementation class in the `def settings_class(self) -> Type[StackComponentSettings]` property.
```

## Tips and best practices

> * The implementation class is loaded and resolved by using the zenml repository root (i.e. the place where the `.zen` folder lives via `zenml init`). Therefore, every one on the team should remember to execute `zenml init` in a consistent manner (usually at the root of the repository where the `.git` folder lives). If the `zenml init` command is not executed, the current working directory is used, which could leave to unexpected results.

Reviewer 2: ```suggestion
* The implementation class is loaded and resolved by using the zenml repository root (i.e. the place where the `.zen` folder lives via `zenml init`). Therefore, every one on the team should remember to execute `zenml init` in a consistent manner (usually at the root of the repository where the `.git` folder lives). If the `zenml init` command is not executed, the current working directory is used, which could lead to unexpected results.
```

---
description: How to extend ZenML with plug-in flavor components.
---

> # Develop a Custom Component Flavor in ZenML

Reviewer 3: Wonder if we should have some language which is a bit more open. Something like "use your own tools with ZenML" or something. Maybe out of place in this kind of documentation, but it might help someone to know at a quick glance what this page is about, or how to find what they're looking for when they don't know what component or flavor means in the context of ZenML.

Reviewer 3: Can be in title or in subtitle or maybe just in how we link to the page?

Reviewer 4: agree with Alex here, the title might not be immediatley obvious to newbees

```python
from abc import ABC, abstractmethod
from typing import Type

from zenml.enums import StackComponentType
from zenml.stack import StackComponent, StackComponentConfig, Flavor

class BaseFlavorConfig(StackComponentConfig):
    """Base class for all ZenML flavor configurations.
    
    This is a pydantic class, so pydantic supported items can be
    defined here.
    """

class BaseFlavor(Flavor, ABC):
    """Base class for all ZenML flavors."""

    @property
    def name(self) -> str:
        """The flavor name.

        Returns:
            The flavor name.
        """

    @property
    def docs_url(self) -> Optional[str]:
        """A url to point at docs explaining this flavor.

        Returns:
            A flavor docs url.
        """
        return None

    @property
    def sdk_docs_url(self) -> Optional[str]:
        """A url to point at SDK docs explaining this flavor.

        Returns:
            A flavor SDK docs url.
        """

>         return None

Reviewer 4: Are these relevant for user created flavors? If not we can leave them out in the docs

Reviewer 2: I would say we can even leave out this and the code sample below out completely? They'll get outdated anyway, and people implementing it will import/inspect the file in their IDE anyway and won't get much value out of seeing it here?

Author: @schustmi @AlexejPenner I removed the url methods. But Im a bit confused: Do you mean to leave out:

```
See the full code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/flavor.py#L29).

See the full code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/stack_component.py).
```

If yes then i would push back a bit... the link is important and it can change but the file wont change probably

Reviewer 2: IMO I would keep the link, but remove all the code.

Reviewer 5: Following this discussion I also took out `docs_url`. I didn't fully get which code section you proposed to remove @schustmi @AlexejPenner, `class StackComponent: ...` this one below? Or the code right here? Or both?

    @property
    def logo_url(self) -> Optional[str]:
        """A url to represent the flavor in the dashboard.

        Returns:
            The flavor logo.
        """
        return None

    @property
    @abstractmethod
    def type(self) -> StackComponentType:
        """The stack component type.

        Returns:
            The stack component type.
        """

    # IMPORTANT: This point to the implementation class
    @property
    def implementation_class(self) -> Type[StackComponent]:
        """Implementation class for this flavor.

        Returns:
            The implementation class for this flavor.
        """

    @property
    def config_class(self) -> Type[StackComponentConfig]:
        """Returns `StackComponentConfig` config class.

        Returns:
            The config class.
        """
```

> See the code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/flavor.py#L29).

Reviewer 3: ```suggestion
See the full code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/flavor.py#L29).
```

        """Cleans up resources after the step run is finished.

        Args:
            info: Info about the step that was executed.
            step_failed: Whether the step failed.
        """
```

See the full code [here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/stack_component.py).

{% hint style="info" %}

In ZenML, a flavor's `config` and `settings` are two separate, yet related entities. The `config` is the static part of your flavor's configuration, defined when you register your flavor. The `settings` are the dynamic part of your flavor's configuration, optionally defined when you register your flavor but can be overridden at runtime. You can read more about the differences [here](../../user-guide/advanced-guide/pipelining-features/configure-steps-pipelines.md).

{% endhint %}

As each component might define a different interface, please take a look at each component type to read further details on how to define a custom component of that type. For example, read the [step operator](step-operators/custom.md) guide to learn how to create a custom step operator.

## Registering a Flavor

> Once you have implemented your custom flavor and implementation class, you can register your component using the ZenML CLI. Use the `flavor register` command and provide the path to your flavor class:

Reviewer 3: ```suggestion
Once you have implemented your custom flavor and implementation class, you can register your component using the ZenML CLI. Use the `flavor register` command and provide the path to your flavor class (see further below for a concrete example):
```

For example, if your custom orchestrator flavor class `MyOrchestratorFlavor` is defined in `flavors/my_flavor.py`, you'd register it by doing:

```shell
zenml orchestrator flavor register flavors.my_flavor.MyOrchestratorFlavor
```

### Listing Flavors

You can list all registered flavors for a component type using the `flavor list` command:

```shell
zenml <component_type> flavor list
```

Replace `<component_type>` with the type of component you're interested in, e.g. `orchestrator`.

### Understanding Flavor Registration

When you register a flavor, ZenML imports the flavor class and stores its configuration in the ZenML repository. This allows ZenML to use your flavor when running pipelines, even if the dependencies required by the flavor are not installed in the current environment.

In ZenML, the config is defined in flavor class in the `def config_class(self) -> Type[StackComponentConfig]` method, while in the settings are defined in the implementation class in the `def settings_class(self) -> Type[StackComponentSettings]` function.

> ## Tips for Developing Flavors

Reviewer 4: What I think is important here is to talk where the code is loaded from and how to make sure the whole team has access to it.

Author: I added one bullet, can you review @AlexejPenner

Reviewer 5: Rewrote it a bit but looks good IMO so I'll resolve this

> * You can keep changing the `Config` and `Settings` of your flavor after registration. ZenML will pick up these "live" changes when running pipelines.

Reviewer 2: Maybe we should note that changing the config in a breaking way requires an update of the component (not flavor)?

E.g., adding a mandatory `name` to flavor X field will break a registered component of that flavor. Not sure if our CLI `update` commands work in this case or if this leads to a completely broken state where users would have to delete and re-register instead

Author: Im not sure how to word this. I added one bullet. can you review and suggest changes @schustmi

* Always test your flavor thoroughly before using it in production. Make sure it works as expected and handles errors gracefully.

* Keep your flavor code clean and well-documented. This will make it easier for others to use and contribute to your flavor.

* Follow best practices for the language and libraries you're using. This will help ensure your flavor is efficient, reliable, and easy to maintain.

> Probably the best way to learn about developing flavors it to see the code for the [officially supported flavors of ZenML](https://github.com/zenml-io/zenml/tree/main/src/zenml/integrations).

Reviewer 3: ```suggestion
Probably the best way to learn about developing flavors is to see the code for the [officially supported flavors of ZenML](https://github.com/zenml-io/zenml/tree/main/src/zenml/integrations).
```
